core/src/trezor/wire/thp/writer.py

```diff
@@ -0,0 +1,100 @@
+from micropython import const
+from trezorcrypto import crc
+from typing import TYPE_CHECKING
+
+from trezor import io, loop, utils
+
+from . import PacketHeader
+
+INIT_HEADER_LENGTH = const(5)
+CONT_HEADER_LENGTH = const(3)
+CHECKSUM_LENGTH = const(4)
+MAX_PAYLOAD_LEN = const(60000)
+MESSAGE_TYPE_LENGTH = const(2)
+
+if TYPE_CHECKING:
+    from trezorio import WireInterface
+    from typing import Awaitable, Sequence
+
+if __debug__:
+    from trezor import log
+
+
+def write_payload_to_wire_and_add_checksum(
+    iface: WireInterface, header: PacketHeader, transport_payload: bytes
+) -> Awaitable[None]:
+    header_checksum: int = crc.crc32(header.to_bytes())
+    checksum: bytes = crc.crc32(transport_payload, header_checksum).to_bytes(
+        CHECKSUM_LENGTH, "big"
+    )
+    data = (transport_payload, checksum)
+    return write_payloads_to_wire(iface, header, data)
+
+
+async def write_payloads_to_wire(
+    iface: WireInterface, header: PacketHeader, data: Sequence[bytes]
+) -> None:
+    n_of_data = len(data)
+    total_length = sum(len(item) for item in data)
+
+    current_data_idx = 0
+    current_data_offset = 0
+
+    packet = bytearray(iface.TX_PACKET_LEN)
+    header.pack_to_init_buffer(packet)
+    packet_offset: int = INIT_HEADER_LENGTH
+    packet_number = 0
+    nwritten = 0
+    while nwritten < total_length:
+        if packet_number == 1:
+            header.pack_to_cont_buffer(packet)
+        if packet_number >= 1 and nwritten >= total_length - iface.TX_PACKET_LEN:
+            packet[:] = bytearray(iface.TX_PACKET_LEN)
+            header.pack_to_cont_buffer(packet)
+        while True:
+            n = utils.memcpy(
+                packet, packet_offset, data[current_data_idx], current_data_offset
+            )
+            packet_offset += n
+            current_data_offset += n
+            nwritten += n
+
+            if packet_offset < iface.TX_PACKET_LEN:
+                current_data_idx += 1
+                current_data_offset = 0
+                if current_data_idx >= n_of_data:
+                    break
+            elif packet_offset == iface.TX_PACKET_LEN:
+                break
+            else:
+                raise Exception("Should not happen!!!")
+        packet_number += 1
+        packet_offset = CONT_HEADER_LENGTH
+
+        # write packet to wire (in-lined)
+        if __debug__:
+            log.debug(
+                __name__,
+                "write_packet_to_wire: %s",
+                utils.hexlify_if_bytes(packet),
+                iface=iface,
+            )
+        written_by_iface: int = 0
+        while written_by_iface < len(packet):
+            await loop.wait(iface.iface_num() | io.POLL_WRITE)
+            written_by_iface = iface.write(packet)
```

Comment:
Do we support partially sent packets? (here and below)
If so:

```suggestion
            written_by_iface += iface.write(packet)
```

Reply:
This code is based on the current implementation of `codec_v1` in `main`.
https://github.com/trezor/trezor-firmware/blob/566f6fe7cbdbc3b23c872de05b6314d89e931f1c/core/src/trezor/wire/codec/codec_v1.py#L100-L104

@matejcik - what is your opinion on that - do we need the `written_by_iface` at all? (note: it is denoted `n` in the codec_v1 implementation)

Reply:
This code is based on the current implementation of `codec_v1` in `main`.
https://github.com/trezor/trezor-firmware/blob/566f6fe7cbdbc3b23c872de05b6314d89e931f1c/core/src/trezor/wire/codec/codec_v1.py#L100-L104

@matejcik - what is your opinion on that - do we need the `written_by_iface` at all? (note: it is denoted `n` in the codec_v1 implementation)

Reply:
> None of our transports supports partial writes, I think.

So do I. But I am wondering if there is some reason behind the design, as it was like that for a *long* time.

Reply:
well, the write can return either `len(report)`, or `0` -- or some weird other value which we aren't ready to handle.
i'd spell the loop (even the current one) slightly differently:
```
while True:
    n = iface.write(packet)
    if n == len(packet):
        break
    assert n == 0
```
to answer the question, we do want to check for a zero here

Reply:
Note: for this particular case, I have looked at the implementation for stm32 webusb. There are a few error cases that can potentially return value `-1` (I am not sure if it can happen in real-life situation.) Code snippet:
```
int usb_webusb_write(uint8_t iface_num, const uint8_t *buf, uint32_t len) {
  volatile usb_webusb_state_t *state = usb_get_webusb_state(iface_num);
  if (state == NULL) {
    return -1;  // Invalid interface number
  }

  if (state->dev_handle == NULL) {
    return -1;  // Class driver not initialized
  }
  
  state->ep_in_is_idle = 0;
  USBD_LL_Transmit(state->dev_handle, state->ep_in, UNCONST(buf),
                   (uint16_t)len);

  return len;
```
Should we somehow handle` -1 `too, or is failing on `assert n == 0` enough?

Reply:
IIUC, -1 is used to indicate an error, so it probably should be converted into an exception here:
https://github.com/trezor/trezor-firmware/blob/0c2ff24398b8b44740b05cc1bb9af821e309fc84/core/embed/upymod/modtrezorio/modtrezorio-webusb.h#L124-L125
(similarly to how it's done in `mod_trezorio_WebUSB_read`)

Reply:
> IIUC, -1 is used to indicate an error, so it probably should be converted into an exception [...]

This could be quite possibly the case...

Just to clarify, I am mostly asking  "why do we want to have check for 0" and "why do we should use `assert` over `utils.ensure`"  - to get a better general understanding of what-to-check (why check the 0 here) and how-to-check (assert vs utils.ensure) :)

Reply:
> Just to clarify, I am mostly asking "why do we want to have check for 0" and "why do we should use `assert` over `utils.ensure`" - to get a better general understanding of what-to-check (why check the 0 here) and how-to-check (assert vs utils.ensure) :)

@matejcik ?